File: plugins/collector/event/apps/database/postgre.go
package database

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type PostgreSql struct {
	version string
	reg     *regexp.Regexp
}

func (PostgreSql) Name() string { return "postgresql" }

func (PostgreSql) Type() string { return "database" }

func (p PostgreSql) Version() string { return p.version }

func (PostgreSql) Match(proc *process.Process) bool {
	return proc.Name == "postgres" && strings.Contains(proc.Argv, "config_file")
}

func (p *PostgreSql) Run(proc *process.Process) (mapping map[string]string, err error) {
	if p.reg == nil {
		p.reg = regexp.MustCompile(`(\d+\.)+\d+`)
	}
	result, err := apps.Execute(proc, "-V")
	if err != nil {
		return nil, err
	}
	str := p.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	p.version = str
	return
}

func init() {
	apps.Regist(&PostgreSql{})
}

File: plugins/collector/src/event/apps/database/postgre.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct PostgreSql {
    version: String,
}

impl PostgreSql {
    pub fn new() -> Self { PostgreSql { version: String::new() } }
}

impl IApp for PostgreSql {
    fn name(&self)     -> &'static str { "postgresql" }
    fn app_type(&self) -> &'static str { "database" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool {
        p.name == "postgres" && p.argv.contains("config_file")
    }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["-V"])?;
        // "postgres (PostgreSQL) 14.5"
        self.version = find_version(&out);
        if self.version.is_empty() {
            return Err(anyhow::anyhow!("version not found"));
        }
        Ok(HashMap::new())
    }
}
